sklearn/gaussian_process/kernels.py

```diff
@@ -1617,6 +1617,132 @@ def __repr__(self):
             self.__class__.__name__, self.length_scale, self.periodicity)
 
 
+class LinearKernel(Kernel):
+    """Linear kernel.
+
+    The Linear kernel is non-stationary and can be obtained from Bayesian
+    linear regression by shifting the origin by c. The parameters of the Linear
+    kernel are about specifying the origin. The kernel is given by:
+
+    k(x_i, x_j) = ((x_i - c) \cdot (x_j - c))
+
+    The Linear kernel can be combined with other kernels, more commonly with
+    periodic kernels.
+
+    Parameters
+    ----------
+    c : float or array with shape (n_features,), default: 0.0
+        The offset for the origin. If a float, the same offset is used
+        throughout all dimensions. If an array, each dimension of c defines the
+        offset of the respective feature dimension.
+
+    c_bounds : pair of floats >= 0, default: (1e-5, 1e5)
+        The lower and upper bound on c
+    """
+    def __init__(self, c=0.0, c_bounds=(1e-5, 1e5)):
+        self.c = c
+        self.c_bounds = c_bounds
+
+    @property
+    def non_uniform_offset(self):
+        return np.iterable(self.c) and len(self.c) > 1
+
+    @property
+    def hyperparameter_c(self):
+        if self.non_uniform_offset:
+            return Hyperparameter("c", "numeric", self.c_bounds, len(self.c))
+        else:
+            return Hyperparameter("c", "numeric", self.c_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """Return the kernel k(X, Y) and optionally its gradient.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples_X, n_features)
+            Left argument of the returned kernel k(X, Y)
+
+        Y : array, shape (n_samples_Y, n_features), (optional, default=None)
+            Right argument of the returned kernel k(X, Y). If None, k(X, X)
+            if evaluated instead.
+
+        eval_gradient : bool (optional, default=False)
+            Determines whether the gradient with respect to the kernel
+            hyperparameter is determined. Only supported when Y is None.
+
+        Returns
+        -------
+        K : array, shape (n_samples_X, n_samples_Y)
+            Kernel k(X, Y)
+
+        K_gradient : array (opt.), shape (n_samples_X, n_samples_X, n_dims)
+            The gradient of the kernel k(X, X) with respect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+        X = np.atleast_2d(X)
+        self.c = _check_length_scale(X, self.c)
+        if Y is None:
+            K = np.inner(X - self.c, X - self.c)
+        else:
+            if eval_gradient:
+                raise ValueError(
+                    "Gradient can only be evaluated when Y is None.")
+            K = np.inner(X - self.c, Y - self.c)
+
+        if eval_gradient:
+            if self.hyperparameter_c.fixed:
+                c_gradient = np.empty((X.shape[0], X.shape[0], 0))
```

Comment:
empty means it will be filled with junk values ... the memory will not be cleared. I doubt this is what you want. Is it not sufficiently tested?

But I also wonder if we can avoid allocating this large array at all in the fixed c case. Can we just do the equivalent of `np.broadcast_to(0, (X.shape[0], X.shape[0]))`? (I say the equivalent, because broadcast_to is only available from numpy 1.10.) Whether this is worthwhile depends on how the gradient matrix is then used. Will the memory be materialised anyway? Does the gradient need to not be read-only?

Reply:
Hmmm. I did not write this portion myself, but that does seem unusual. I did notice that this practice seems to be ubiquitous amongst the kernels when no gradient is evaluated. It could be that the values are not used when a parameter is `fixed`. 

I'll look into this a bit more.

Reply:
The testing is of generic properties of a kernel, not of the specific kernel.

Yes, please look into these `empty` calls... It's weird to allocate an array randomly. Either we don't need the array or we need it to contain something meaningful.

Reply:
The array has a 0 dimension, so they don't appear to be populated with anything (I did not know that was possible). Their size is 0. and `sys.getsizeof()` reports it to be 128 bytes (0 bytes from `.nbytes`), perhaps due to the size information it holds. The size seems to be independent of the size of `X`.

The gradient is deeply integrated into the code, so removing it outright isn't practical. I replaced it with an empty array of size (0,) (96 bytes) which works in all cases except for kernel addition where the following exception is raised:

`ValueError: all the input array dimensions except for the concatenation axis must match exactly`

No shape appears to work for kernel addition except `(X.shape[0], X.shape[0], 0)`. For all other kernel operations, the shape of the matrix has no effect.